src/access/AccessRestriction.h

```diff
@@ -0,0 +1,288 @@
+ */
+
+#pragma once
+
+#include "RequestPath.h"
+#include "SubjectDescriptor.h"
+#include <algorithm>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <cstdint>
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/Optional.h>
+#include <lib/support/CHIPMem.h>
+#include <map>
+#include <memory>
+#include <protocols/interaction_model/Constants.h>
+#include <vector>
+
+namespace chip {
+namespace Access {
+
+template <typename T>
+using SharedPtr = chip::Platform::SharedPtr<T>;
+
+class AccessRestriction
+{
+public:
+    static constexpr size_t kNumberOfFabrics      = CHIP_CONFIG_MAX_FABRICS;
+    static constexpr size_t kEntriesPerFabric     = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_ENTRIES_PER_FABRIC;
+    static constexpr size_t kRestrictionsPerEntry = CHIP_CONFIG_ACCESS_RESTRICTION_MAX_RESTRICTIONS_PER_ENTRY;
+
+    /**
+     * Defines the type of access restriction, which is used to determine the meaning of the restriction's id.
+     */
+    enum class Type : uint8_t
+    {
+        kAttributeAccessForbidden = 0,
+        kAttributeWriteForbidden  = 1,
+        kCommandForbidden         = 2,
+        kEventForbidden           = 3
+    };
+
+    /**
+     * Defines a single restriction on an attribute, command, or event.
+     *
+     * If id is not set, the restriction applies to all attributes, commands, or events of the given type (wildcard).
+     */
+    struct Restriction
+    {
+        Type restrictionType;
+        Optional<uint32_t> id;
+    };
+
+    /**
+     * Defines a single entry in the access restriction list, which contains a list of restrictions
+     * for a cluster on an endpoint.
+     */
+    struct Entry
+    {
+        FabricIndex fabricIndex;
+        EndpointId endpointNumber;
+        ClusterId clusterId;
+        std::vector<Restriction> restrictions;
+
+        bool IsValid() const
+        {
+            return endpointNumber != 0 && clusterId != app::Clusters::NetworkCommissioning::Id &&
+                clusterId != app::Clusters::Descriptor::Id;
+        }
+    };
+
+    /**
+     * Used to notify of changes in the access restriction list and active reviews.
+     */
+    class EntryListener
+    {
+    public:
+        enum class ChangeType
+        {
+            kAdded   = 1,
+            kRemoved = 2,
+            kUpdated = 3
+        };
+
+        virtual ~EntryListener() = default;
+
+        /**
+         * Notifies of a change in the access restriction list.
+         *
+         * @param [in] fabricIndex  The index of the fabric in which the entry has changed.
+         * @param [in] index        Index of entry to which has changed (relative to fabric).
+         * @param [in] entry        The latest value of the entry which changed.
+         * @param [in] changeType   The type of change that occurred.
+         */
+        virtual void OnEntryChanged(FabricIndex fabricIndex, size_t index, SharedPtr<Entry> entry, ChangeType changeType) = 0;
```

Comment:
I assume this leads to the cluster event getting triggered?

Reply:
Yes.  access-control-server is a listener and generates the event based on this callback.

Reply:
This is missing `MatterReportingAttributeChangeCallback` in the impl of the storage.

Reply:
Added to access-control-server.cpp.